snuba/datasets/plans/translator/mapper.py

```diff
@@ -0,0 +1,50 @@
+from abc import ABC, abstractmethod
+from typing import Generic, Optional, Sequence, TypeVar
+
+TExpIn = TypeVar("TExpIn")
+TExpOut = TypeVar("TExpOut")
+# The translator to be recursively applied to the children of an expression
+TTranslator = TypeVar("TTranslator")
+
+
+class ExpressionMapper(
+    ABC, Generic[TExpIn, TExpOut, TTranslator],
+):
+    """
+    One translation rule used by the a mapping expression translator to
+    translate an expression.
+
+    An ExpressionMapper is supposed to know how to map an expression but
+    it is not required to map its children. For that it receives a
+    translator of TTranslator type.
```

Comment:
Some additional context on why we need to pass the translator to an ExpressionMapper and delegate to the mapper the responsibility to apply the translator to its children:

in the original AST transform method, the FunctionCall node guides this process: it first applies the transformation lambda to all children (parameters), then it replaces the old parameter with the new one on the FunctionaCall node and then it applies the lambda to the new node and returns the result.

I originally wanted to do something like this where the visitor ensures that first we translate the parameter of a function then we find a matcher for the function itself with the translated parameters.
Unfortunately the translation produces a different data type than the one it takes in input (they are not Expressions since they are the equivalent of the clickhouse query) so that simple process breaks. We cannot pack the logical pre-translation FunctionCall with the post-translation parameters and find a mapper for that because it would not be a valid expression.
Thus we need the mapper to return a fully translated subtree (that type checks properly), and we provide a translator to the mapper so it does not have to figure out how to translate its children.